intro_to_ada.rst

```diff
@@ -1904,23 +1904,747 @@ Abstract data types
+
+.. code-block:: ada
+
+    with Ada.Text_IO;
+    use  Ada.Text_IO;
+
+    with Simple_Generic_Array_Pkg;
+
+    procedure Test_Reverse_Colors_Simple_Pkg is
+
+       generic
+          type T is private;
+          with package P is new Simple_Generic_Array_Pkg (T => T, Index => Integer);
+       procedure Reverse_Array (X : in out P.Array_T);
+
+       procedure Reverse_Array (X : in out P.Array_T) is
+       begin
+          for I in X'First .. (X'Last + X'First) / 2 loop
+             declare
+                Tmp     : T;
+                X_Left  : T renames X (I);
+                X_Right : T renames X (X'Last + X'First - I);
+             begin
+                Tmp     := X_Left;
+                X_Left  := X_Right;
+                X_Right := Tmp;
+             end;
+          end loop;
+       end Reverse_Array;
+
+       type Color is (Black, Red, Green, Blue, White);
+
+       package Color_Pkg is new Simple_Generic_Array_Pkg (T => Color, Index => Integer);
+
+       procedure Reverse_Color_Array is new Reverse_Array (T => Color, P => Color_Pkg);
+
+       My_Colors : Color_Pkg.Array_T (1 .. 5) := (Black, Red, Green, Blue, White);
+    begin
+       for C of My_Colors loop
+          Put_Line ("My_Color: " & Color'Image (C));
+       end loop;
+
+       New_Line;
+       Put_Line ("Reversing My_Color...");
+       New_Line;
+       Reverse_Color_Array (My_Colors);
+
+       for C of My_Colors loop
+          Put_Line ("My_Color: " & Color'Image (C));
+       end loop;
+
+    end Test_Reverse_Colors_Simple_Pkg;
+
+In this example, we're first instantiating the
+``Simple_Generic_Array_Pkg`` package, thereby creating the ``Color_Pkg``
+package. We then proceed to use this ``Color_Pkg`` package in the
+instantiation of the generic ``Reverse_Array`` procedure. Also, in the
+declaration of the ``My_Colors`` array, we make use of the array type
+definition from the ``Color_Pkg`` package.
+
+Abstracting procedures into packages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In the previous example, we moved the array type definition into a
+separate package, but leaved the generic procedure (``Reverse_Array``) in
+the test application. Another approach would have been to also move the
+generic procedure into the generic package. The advantage of this approach
+is that we don't need to repeat the formal declaration for the
+``Reverse_Array`` procedure. Also, this simplifies the instantiation in
+the test application.
+
+First, we need to extend the previous package by adding the declaration of
+the ``Reverse_Array`` procedure. Note that we've just renamed the
+``Simple_Generic_Array_Pkg`` package to ``Generic_Array_Pkg`` in order to
+avoid confusion with the previous example. This is the resulting
+specification:
+
+.. code-block:: ada
+
+    generic
+       type T is private;
+       type Index is range <>;
+    package Generic_Array_Pkg is
+       type Array_T is array (Index range <>) of T;
+
+       procedure Reverse_Array (X : in out Array_T);
```

Comment:
This approach has a lot of problems in terms of size of code generated (speaking from experience). Having this nice array package with utils that works on array on it will encourage people to use that by convenience.

But the result, with GNAT, will be an explosion of compilation time and ultimately code size.

So I believe that the previous approach that you showcased for this particular example, to have a separate generic function, was much betteR.

Also, and this is maybe more important: This section is *basic* generics, so it should come before the section on formal packages. In my opinion, rather than trying to keep the same example here, you should showcase a different example of an abstract data type (like, for example the stack type, or something similar), where it really makes sense to put all of that in a generic package. This will probably include a private part.

Reply:
I'm surprised that this approach cannot be recommended. Isn't this something that should be improved in the compiler?

In any case, I'd rather keep the example and mention the problem with the generated code as an alert to the programmer.